Simple load-store program does not terminate (probably LSQ related)

Given the following C program:
```c
#include "dynamatic/Integration.h"
#include <stdint.h>

uint32_t test27(uint16_t var1[16],
           float var5[16], int16_t var12[32]) {
  var1[0] = var12[0];
  var1[5] = var12[0];
  var1[0 + var5[0] > 5] = 0;
  return 0;
}

void test_bench() {
  uint16_t var1[16] = {};
  float var5[16] = {};
  int16_t var12[32] = {};
  CALL_KERNEL(test27, var1, var5, var12);
}
int main() { test_bench(); }
```
Running it with dynamatic leads to seemingly a deadlock (doesn't finish within 20k cycles which should be more than enough).

Script:
```
dynamatic> set-dynamatic-path /home/mboeck/dynamatic
dynamatic> set-src test27.c
dynamatic> compile
[INFO] Compiled to LLVM IR
[INFO] Optimized LLVM IR
[INFO] Applied memory dependency analysis to LLVM IR
[INFO] Converted to std dialect
[INFO] Applied CF transformations
[INFO] Marked memory accesses with the corresponding interfaces in cf
[INFO] Compiled cf to handshake
[INFO] Applied transformations to handshake
[INFO] Running simple buffer placement (on-merges).
[INFO] Placed simple buffers
[INFO] Canonicalized handshake
[INFO] Created test27 DOT
[INFO] Converted test27 DOT to PNG
[INFO] Created test27_CFG DOT
[INFO] Converted test27_CFG DOT to PNG
[INFO] Lowered to HW
[INFO] Compilation succeeded
dynamatic> write-hdl
[INFO] Exported RTL (vhdl)
[INFO] HDL generation succeeded
dynamatic> simulate --timeout 20000
[INFO] Built kernel for IO gen.
[INFO] Ran kernel for IO gen.
[INFO] Launching simulation (vsim)
[FATAL] Simulation failed
```

Report output:
```
# ** Warning: There is an 'U'|'X'|'W'|'Z'|'-' in an arithmetic operand, the result will be 'X'(es).
#    Time: 22 ns  Iteration: 4  Instance: /tb/duv_inst/addf0/operator/roundingAdder
# ** Warning: There is an 'U'|'X'|'W'|'Z'|'-' in an arithmetic operand, the result will be 'X'(es).
#    Time: 22 ns  Iteration: 4  Instance: /tb/duv_inst/addf0/operator/roundingAdder
# ** Failure: ERROR: Maximum clock count of 20000 reached; Assuming non-termination
#    Time: 80014 ns  Iteration: 1  Process: /tb/#MERGED#gen_sim_timeout_proc,gen_sim_latency_proc File: /home/mboeck/dynamatic/build/bin/bug89/out/sim/HDL_SRC/tb_test27.vhd
# Break in Architecture behavior at /home/mboeck/dynamatic/build/bin/bug89/out/sim/HDL_SRC/tb_test27.vhd line 478
# Stopped at /home/mboeck/dynamatic/build/bin/bug89/out/sim/HDL_SRC/tb_test27.vhd line 478
# End time: 16:02:30 on May 06,2026, Elapsed time: 0:00:02
# Errors: 1, Warnings: 184
===-------------------------------------------------------------------------===
                         Miscellaneous Ungrouped Timers
===-------------------------------------------------------------------------===

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.0000 (100.0%)   0.0004 (100.0%)   0.0004 (100.0%)  14.0773 (100.0%)  Simulator runtime
   0.0000 (100.0%)   0.0004 (100.0%)   0.0004 (100.0%)  14.0773 (100.0%)  Total

[ERROR COMPARE] Token mismatch: [0x00000000] and [[[[runtime]]]] are not equal (at transaction id 0).
Comparison of [var1] : Fail
[ERROR HLS_VERIFIER] C and VHDL outputs do not match
```

Commit tested: https://github.com/EPFL-LAP/dynamatic/commit/6a891fa910a3dc72838c7adccc82ff10ac57c0bb

BTW, I got this from the profiler (maybe you didn't enable MILP buffer placement

```
/data/dhls-utils/../dynamatic/integration-test/test0/test0.c:17:5: error:
      constexpr function never produces a constant expression [-Winvalid-constexpr]
   17 |     test_bench() {
      |     ^
/data/dhls-utils/../dynamatic/integration-test/test0/test0.c:21:3: note:
      non-constexpr function 'callKernel<unsigned int, unsigned short *, float *, short *,
      unsigned short (&)[16], float (&)[16], short (&)[32]>' cannot be used in a constant
      expression
   21 |   CALL_KERNEL(test0, var1, var5, var12);
      |   ^
/data/dhls-utils/../dynamatic/include/dynamatic/Integration.h:225:39: note:
      expanded from macro 'CALL_KERNEL'
  225 | #define CALL_KERNEL(kernelAndArgs...) callKernel(kernelAndArgs)
      |                                       ^
/data/dhls-utils/../dynamatic/include/dynamatic/Integration.h:220:12: note:
      declared here
  220 | static Res callKernel(Res (*kernel)(FunArgs...), RealArgs &&...args) {
      |            ^
1 error generated.


```

> BTW, I got this from the profiler (maybe you didn't enable MILP buffer placement
> 
> ```
> /data/dhls-utils/../dynamatic/integration-test/test0/test0.c:17:5: error:
>       constexpr function never produces a constant expression [-Winvalid-constexpr]
>    17 |     test_bench() {
>       |     ^
> /data/dhls-utils/../dynamatic/integration-test/test0/test0.c:21:3: note:
>       non-constexpr function 'callKernel<unsigned int, unsigned short *, float *, short *,
>       unsigned short (&)[16], float (&)[16], short (&)[32]>' cannot be used in a constant
>       expression
>    21 |   CALL_KERNEL(test0, var1, var5, var12);
>       |   ^
> /data/dhls-utils/../dynamatic/include/dynamatic/Integration.h:225:39: note:
>       expanded from macro 'CALL_KERNEL'
>   225 | #define CALL_KERNEL(kernelAndArgs...) callKernel(kernelAndArgs)
>       |                                       ^
> /data/dhls-utils/../dynamatic/include/dynamatic/Integration.h:220:12: note:
>       declared here
>   220 | static Res callKernel(Res (*kernel)(FunArgs...), RealArgs &&...args) {
>       |            ^
> 1 error generated.
> ```

Yes I use the default buffer placement for compile time reasons.
You can remove the `constexpr` and surrounding preprocessor macros for reproduction. I've edited the issue

--- a/tools/backend/lsq-generator-python/lsq-generator.py
+++ b/tools/backend/lsq-generator-python/lsq-generator.py
@@ -406,7 +406,7 @@ class LSQWrapper:
         # wreq_ready, wresp_valid, wresp_id
         self.lsq_wrapper_str += "\t----------------------------------------------------------------------------\n"
         self.lsq_wrapper_str += (
-            "\t-- Process for wreq_ready, wresp_valid and wresp_id (1)\n"
+            "\t-- Process for wreq_ready, wresp_valid and wresp_id\n"
         )
         self.lsq_wrapper_str += self.reg_init_str
         self.lsq_wrapper_str += "\t" * \
@@ -427,12 +427,12 @@ class LSQWrapper:
         for i in range(self.lsq_config.numStMem):
             self.lsq_wrapper_str += OpTab(wreq_ready[i], (self.tab_level + 2), "'1'")
 
-
         self.lsq_wrapper_str += (
             "\n"
             + "\t" * (self.tab_level + 2)
             + "if "
-            + io_storeEn.getNameWrite() + " = '1' then\n"
+            + io_storeEn.getNameWrite()
+            + " = '1' then\n"
         )
 
         for i in range(self.lsq_config.numStMem):

--- a/tools/backend/lsq-generator-python/lsq-generator.py
+++ b/tools/backend/lsq-generator-python/lsq-generator.py
@@ -428,13 +428,11 @@ class LSQWrapper:
             self.lsq_wrapper_str += OpTab(wreq_ready[i], (self.tab_level + 2), "'1'")
 
 
-        self.lsq_wrapper_str += "\t-- Signals the LSQ core that the store is completed if the MC is ready.\n"
-        self.lsq_wrapper_str += "\t-- NOTE: we assume that MC joins the addr and data, so here we only check the addr ready .\n"
         self.lsq_wrapper_str += (
             "\n"
             + "\t" * (self.tab_level + 2)
             + "if "
-            + io_storeEn.getNameWrite() + " = '1' " + " and " + io_stAddrToMC_ready.getNameRead() + " = '1' then\n"
+            + io_storeEn.getNameWrite() + " = '1' then\n"
         )
 
         for i in range(self.lsq_config.numStMem):

--- a/tools/backend/lsq-generator-python/lsq-generator.py
+++ b/tools/backend/lsq-generator-python/lsq-generator.py
@@ -896,7 +896,7 @@ class LSQWrapper:
         )
 
         self.lsq_wrapper_str += "\t-- Signals the LSQ core that the store is completed if the MC is ready.\n"
-        self.lsq_wrapper_str += "\t-- NOTE: we assume that MC joins the addr and data, so here we only check the addr ready .\n"
+        self.lsq_wrapper_str += "\t-- NOTE: we assume that MC joins the addr and data, so here we only check the addr ready.\n"
         self.lsq_wrapper_str += (
             "\n"
             + "\t" * (self.tab_level + 2)

--- a/tools/backend/lsq-generator-python/lsq-generator.py
+++ b/tools/backend/lsq-generator-python/lsq-generator.py
@@ -406,7 +406,7 @@ class LSQWrapper:
         # wreq_ready, wresp_valid, wresp_id
         self.lsq_wrapper_str += "\t----------------------------------------------------------------------------\n"
         self.lsq_wrapper_str += (
-            "\t-- Process for wreq_ready, wresp_valid and wresp_id\n"
+            "\t-- Process for wreq_ready, wresp_valid and wresp_id (1)\n"
         )
         self.lsq_wrapper_str += self.reg_init_str
         self.lsq_wrapper_str += "\t" * \
@@ -427,12 +427,14 @@ class LSQWrapper:
         for i in range(self.lsq_config.numStMem):
             self.lsq_wrapper_str += OpTab(wreq_ready[i], (self.tab_level + 2), "'1'")
 
+
+        self.lsq_wrapper_str += "\t-- Signals the LSQ core that the store is completed if the MC is ready.\n"
+        self.lsq_wrapper_str += "\t-- NOTE: we assume that MC joins the addr and data, so here we only check the addr ready .\n"
         self.lsq_wrapper_str += (
             "\n"
             + "\t" * (self.tab_level + 2)
             + "if "
-            + io_storeEn.getNameWrite()
-            + " = '1' then\n"
+            + io_storeEn.getNameWrite() + " = '1' " + " and " + io_stAddrToMC_ready.getNameRead() + " = '1' then\n"
         )
 
         for i in range(self.lsq_config.numStMem):
@@ -895,12 +897,14 @@ class LSQWrapper:
             "\t" * (self.tab_level + 1) + "elsif rising_edge(clock) then\n"
         )
 
+        self.lsq_wrapper_str += "\t-- Signals the LSQ core that the store is completed if the MC is ready.\n"
+        self.lsq_wrapper_str += "\t-- NOTE: we assume that MC joins the addr and data, so here we only check the addr ready .\n"
         self.lsq_wrapper_str += (
             "\n"
             + "\t" * (self.tab_level + 2)
             + "if "
             + io_storeEn.getNameWrite()
-            + " = '1' then\n"
+            + " = '1'" + " and " + io_stAddrToMC_ready.getNameRead() + " = '1' then\n"
         )
 
         for i in range(self.lsq_config.numStMem):

